File: spinodal_decomp/FD_solver/laplacian.m
% function arrOut = laplacian(arrIn)
% % Calculates laplacian for a matrix using a 3x3 convolution with edge wrapping
% arrOut = -arrIn + ...
%     0.2(circshift(arrIn,1)+circshift(arrIn,-1)+circshift(arrIn,-1,2)+circshift(arrIn,1,2));
%     %  + ...
%     % 0.05*(circshift(arrIn,[1 1])+circshift(arrIn,[1 -1])+circshift(arrIn,[-1 1])+circshift(arrIn,[-1 -1]));
% end

function arrOut = laplacian(arrIn)
    [nx,ny] = size(arrIn);

    arrOut = zeros(nx,ny); %Initialize Laplacian
    for i = 1:nx
        for j = 1:ny
            if i > 1
                dadx_L = arrIn(i,j)-arrIn(i-1,j);
            else
                dadx_L = 0;
            end
            if i < nx
                dadx_R = arrIn(i+1,j)-arrIn(i,j);
            else
                dadx_R = 0;
            end
            if j > 1
                dady_B = arrIn(i,j)-arrIn(i,j-1);
            else
                dady_B = 0;
            end
            if j < ny
                dady_T = arrIn(i,j+1)-arrIn(i,j);
            else
                dady_T = 0;
            end
            arrOut(i,j) = (dadx_R-dadx_L + dady_T-dady_B);
        end
    end
    
    end

File: spinodal_decomp/CahnHilliard_Python_solvers/CHsolvers/laplacian.py
from . import aux_functions as aux


def laplace(a, nxt, nyt, xright, xleft, yright, yleft, boundary):
    """
    Compute the discrete Laplacian of a
    :param xright:
    :param a: matrix
    :param nxt: nx temp (number of grid points in x-direction, locally defined)
    :param nyt: ny temp (number of grid points in y-direction, locally defined)
    :return: lap_a, the discrete laplacian of a
    """
    lap_a = aux.dmatrix(nxt, nyt)
    h2 = ((xright - xleft) / nxt) ** 2
    for i in range(nxt):
        for j in range(nyt):
            if i > 0:
                dadx_L = a[i, j] - a[i - 1, j]
            else:
                if boundary == "neumann":
                    dadx_L = 0
                elif boundary == "periodic":
                    dadx_L = a[i, j] - a[nxt-1, j]
            if i < nxt - 1:
                dadx_R = a[i + 1, j] - a[i, j]
            else:
                if boundary == "neumann":
                    dadx_R = 0
                elif boundary == "periodic":
                    dadx_R = a[1, j] - a[i, j]
            if j > 0:
                dady_B = a[i, j] - a[i, j - 1]
            else:
                if boundary == "neumann":
                    dady_B = 0
                elif boundary == "periodic":
                    dady_B = a[i, j] - a[i, nyt-1]
            if j < nyt - 1:
                dady_T = a[i, j + 1] - a[i, j]
            else:
                if boundary == "neumann":
                    dady_T = 0
                elif boundary == "periodic":
                    dady_T = a[i, 1] - a[i, j]
            lap_a[i, j] = (dadx_R - dadx_L + dady_T - dady_B) / h2
    return lap_a
